### PR TITLE
Support auto-compounding and cooldown feature.

### PR DESCRIPTION
**Pull request by fb-alexcq**

### CODE REVIEW

**Review by Cphilo — APPROVED** [review]

File: contracts/rewarders/BaseRewarder.sol [hunk 1]
```diff
@@ -59,6 +59,11 @@ abstract contract BaseRewarder is IRewarder, BaseAdminOperation {
         lastRewardBlock = block.number;
     }
 
+    modifier onlyStakingManager() {
+        require(_msgSender() == address(stakingManager), "BaseRewarder: !admin");
```
Comment (merlincdj) [inline contracts/rewarders/BaseRewarder.sol:63]: `"BaseRewarder: !admin"` -> `"BaseRewarder: !stakingManager"`

File: contracts/rewarders/MpRewarder.sol [hunk 2]
```diff
@@ -64,6 +64,11 @@ contract MpRewarder is IRewarder, BaseAdminOperation {
         lastRewardTs = block.timestamp;
     }
 
+    modifier onlyStakingManager() {
+        require(_msgSender() == address(stakingManager), "BaseRewarder: !admin");
```
Comment (merlincdj) [inline contracts/rewarders/MpRewarder.sol:68]: `"BaseRewarder: !admin"` -> `"BaseRewarder: !stakingManager"`

File: contracts/rewarders/BaseRewarder.sol (hunk 1 above)
Comment (fb-alexcq) [reply contracts/rewarders/BaseRewarder.sol:63]: Okay.

File: contracts/WooStakingManager.sol [hunk 3]
```diff
@@ -69,6 +61,8 @@ contract WooStakingManager is IWooStakingManager, BaseAdminOperation {
     IRewarder public mpRewarder; // Record and distribute MP rewards
     EnumerableSet.AddressSet private rewarders; // Other general rewards (e.g. usdc, eth, op, etc)
 
+    IWooStakingCompounder compounder;
```
[the remark this answers is not in the document]
Comment (fb-alexcq) [reply contracts/WooStakingManager.sol:64]: adding `public`

File: contracts/rewarders/MpRewarder.sol [hunk 4]
```diff
@@ -64,6 +64,11 @@ contract MpRewarder is IRewarder, BaseAdminOperation {
         lastRewardTs = block.timestamp;
     }
 
+    modifier onlyStakingManager() {
+        require(_msgSender() == address(stakingManager), "BaseRewarder: !stakingManager");
```
Comment (merlincdj) [inline contracts/rewarders/MpRewarder.sol:68]: `BaseRewarder: !stakingManager` -> `MpRewarder: !stakingManager`

**Review by merlincdj — APPROVED** [review]

File: test/typescript/WooStakingCompounder.test.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+        expect(compounder.removeUser()).to.be.reverted;
+
+        await compounder.setCooldownDuration(5);
+
+        await compounder.removeUsers([user1.address]);
+        expect(await compounder.contains(user1.address)).to.be.eq(false);
+        expect(await compounder.allUsersLength()).to.be.eq(2);
+
+        let allUsers = await compounder.allUsers();
```
Comment (fb-alexcq) [inline test/typescript/WooStakingCompounder.test.ts:74]: add verifying the event please.

File: test/typescript/WooStakingCompounder.test.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        expect(allUsers[0]).to.be.eq(owner.address);
+        expect(allUsers[1]).to.be.eq(user2.address);
+
+        await compounder.compound(0, 1);
+        await compounder.compoundAll();
+    });
```
Comment (fb-alexcq) [inline test/typescript/WooStakingCompounder.test.ts:80]: cooldown logic not tested, right?

Comment (Cphilo) [reply test/typescript/WooStakingCompounder.test.ts:86]: Resolved. Add CoolDown test in StakingManager test.

File: test/typescript/WooStakingCompounder.test.ts (hunk 5 above)
Comment (Cphilo) [reply test/typescript/WooStakingCompounder.test.ts:80]: Resolved

File: test/typescript/WooStakingManager.test.ts [hunk 7]
```diff
@@ -265,6 +269,17 @@ describe("WooStakingManager tests", () => {
         expect(await usdcToken.balanceOf(user2.address)).to.be.eq(user2Bal);
     });
 
+    it("CoolDown Tests", async () => {
+        await stakingCompounder.addUser();
+
+        expect(stakingManager["claimRewards()"]()).to.be.reverted;
+
+        await stakingCompounder.setCooldownDuration(2);
```
Comment (fb-alexcq) [inline test/typescript/WooStakingManager.test.ts:277]: better add a sleep(2) before this line?

Comment (Cphilo) [reply test/typescript/WooStakingManager.test.ts:287]: As discussed, add mine logic. Resolved

File: test/typescript/WooStakingCompounder.test.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            .withArgs(owner.address);
+        await expect(compounder.addUsers([user1.address, user2.address]))
+            .to.emit(compounder, "AddUser")
+            .withArgs(user2.address);
+        let totalUsers = await compounder.allUsersLength();
+        expect(totalUsers).to.be.eq(3);
+
+        await expect(compounder.removeUser()).to.be.revertedWith(
+            "WooStakingCompounder: STILL_IN_COOL_DOWN"
+        );
```
Comment (fb-alexcq) [inline test/typescript/WooStakingCompounder.test.ts:73]: Good!

**Pull request merged**